Globalization support in .NET 5 (#102)
dotnet/designs: accepted/2020/mono-convergence/globalization-support-in-.net5.md

# Globalization Support in .NET 5

## Overview

> Globalization in the .NET Core space has been handled by taking a dependency on the [icu4c](http://site.icu-project.org/home) library and p-invoking into it from managed code.  This model has been successful for desktop and server modes mostly because they can afford to take on the disk space ICU requires.   However, in the size restricted workloads, this is quite a different story as there are differences in platform support and tolerance of taking on a significantly sized dependency.

Reviewer: >  [](start = 176, length = 1)

nit: remove extra space

> Globalization in the .NET Core space has been handled by taking a dependency on the [icu4c](http://site.icu-project.org/home) library and p-invoking into it from managed code.  This model has been successful for desktop and server modes mostly because they can afford to take on the disk space ICU requires.   However, in the size restricted workloads, this is quite a different story as there are differences in platform support and tolerance of taking on a significantly sized dependency.

Reviewer: >dependency on the [icu4c](http://site.icu-project.org/home) [](start = 66, length = 59)

would nice if clarify on Windows we can use Win32 NLS if not using ICU there.

The implementation files which reference globalization interop are:

* [System.Private.CoreLib/src/System/Globalization/IdnMapping.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/IdnMapping.Unix.cs)
* [System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs)
* [System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs)
* [System.Private.CoreLib/src/System/Globalization/TextInfo.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.Unix.cs)
* [System.Private.CoreLib/src/System/Globalization/CompareInfo.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Unix.cs)

> * [System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Unix.cs)

Reviewer: >*  [](start = 0, length = 2)

I believe TimeZoneInfo.Unic.cs is also has dependency on ICU 
[https://github.com/dotnet/runtime/blob/88cc71b26ba1eb430aaeeebddcd5675b3009f2ee/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L127](https://github.com/dotnet/runtime/blob/88cc71b26ba1eb430aaeeebddcd5675b3009f2ee/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L127)

Existing test coverage checks the implementation which is available. It does not have any expected list of data included therefore does not check whether same set of data is available everywhere.

> ## .NET Globalization Invariant Mode

Reviewer 2: The Invariant globalization mode was targeted at building micro-services. It is unusable for apps that display string to non-English speaking users (majority of mobile apps?).

Would it make sense to enhanced the Invariant Mode somewhat (e.g. such that the lower/upper casing works for more than ASCII), without compromising the size-footprint significantly, and make it usable for much broader list of applications?

Reviewer 3: @jkotas I crunched the numbers on this over at https://github.com/dotnet/runtime/issues/30960#issuecomment-535257837. It will increase the size of our RVA statics by just under 3KB to carry the casing data as part of _System.Private.CoreLib.dll_.

Author: I don't have an idea if that's something our customers would accept.  @marek-safar Please chime in here.

Reviewer 4: That makes sense. @steveisok could you add a section about expanding Invariant Mode to handle more than just ASCII.

There is already support for the globalization [invariant mode](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md) since .NET Core 2.0. This will be the most size effective globalization mode for .NET5 which can be set by the developers. We are going to use existing logic and tweak it where necessary to allow IL Linker to remove any globalization code from the final app when running in invariant mode.

## Size Sensitive Platforms

> Existing Xamarin platforms allow customers to choose which set of collation data they would like to add to their apps if any. This still makes sense for .NET5 as on some of the platform ICU dependency is not available from the system and we’ll need to bundle it to the final app. We could make this available to everyone but it’s most likely we will focus only on platforms which are size sensitive in .NET5 time frame.

Reviewer 3: Nit: remove smart quotes from documents that are otherwise ASCII text.

### Android
Android has historically relied on ICU for some kind of globalization support.  From basically the beginning, `Locale` , `Character`, and some classes in `java.text` were provided in the platform globalization.  If you needed more than that, you could access `libicuuc` directly or choose to bundle another version in your APK.  That was pretty much the story up until API level 23.

In API level 24 and higher, the platform white lists its version of `libicuuc` and also includes a subset of [ICU4J](https://developer.android.com/guide/topics/resources/internationalization#relation) API’s.  It’s important to note that [ICU4C](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/) (accessing `libicuuc` directly) is not part of the Android NDK.  What this means is that OEM’s are not required to follow any type of rules and they could limit / alter the data files as they see fit.  The potential for undefined behavior and additional testing on our part is very real. 

Consult this [issue](https://github.com/android/ndk/issues/548#issuecomment-395561629) for detailed commentary on the challenges of making ICU an NDK API.

### What Does This Mean For Us?

> The good news is that p-invoking from managed code works.  Assuming `libicuuc` continues to be white listed, this should not be a problem.  It would be wise if we explored using the `ICU4J`  API subset that’s exposed in the event they decide to tighten things up.  

Reviewer 5: `libicu` is private on iOS/tvOS/watchOS and cannot be used by applications

Reviewer 4: This is android section, ios comments are not relevant here

To summarize:

> * We can p-invoke into the system `libicuuc`.  Globalization for .net5 should work normally

Reviewer 6: What is the difference between `libicu` and `libicuuc`? Do they have the same capabilities and API?

Reviewer: `libicuuc` is the ICU data. `libicu` is the code.

Reviewer: to clarify libicuuc is the common library for accessing the data. https://docs.oracle.com/cd/E36784_01/html/E36873/libicuuc-3lib.html

Reviewer: @danmosemsft where did you see `libicu`?

Reviewer 6: Not in this document. But usually we talk about `libicu` and here I see `libicuuc`. I am naively asking what the difference is.

Reviewer: I believe it meant the same thing. we may be referring to is as libicu for simplicity. the link http://userguide.icu-project.org/packaging is useful for how ICU4C is packaged.

Reviewer 6: ah - thank you

* Android 4.3 (API level 18) started bundling ICU 50 and our minimum version requirement is satisfied. 
* OEM’s can potentially make life difficult by altering locales / globalization features

> * We should explore trying to interop with the `ICU4J` API and determine what the implications, if any are.

Reviewer: >We should explore try [](start = 2, length = 21)

Any idea if anyone buildign ICU packages today that can be used in Android as app local library? I am asking because this can be the solution if the OEMs didn't provide ICU.

Author: It is possible and recommended if you need a version greater than what the particular version of Android supports.  I believe that is less of an issue now.

> ### iOS/tvOS

Reviewer 2: It would be useful to describe what Xamarin does today; and how our customers are happy with it.

Reviewer 5: we ship with the tables that mono includes
we have some, optional `I18N.*.dll`, that customer can opt-in; and
we have some linker tricks to remove a few things (e.g. special calendars when the corresponding I18N support is not included)

Reviewer 5: Customer wise there's not much negative feedback. Either they use Apple API or .net ones.

The most common issue is when mixing both. Apple design is different from .net one, e.g. `EN-fr` is valid on Apple platforms, but does not match to any .net `CultureInfo`. Not very hard to fix, but app developers must be aware of it.

Reviewer 4: >It would be useful to describe what Xamarin does today

Xamarin globalization is not very compatible with .NET Core so this is not that relevant. We can decide iOS/tvOS (and not macOS) be the only platform where the globalization support is very different (and lacking full support for features like System.Globalization.Invariant mode) or unify. The existing implementation is a mix of native and managed pieces which use a subset of ICU data extracted some time ago. It also suffers from the suboptimal performance of string operations as the collation logic is fully managed with no advanced optimizations.

The most-reported bugs we get for globalization support are about inconsistent or missing data

There is no build in support for ICU on apple mobile platforms. For our purposes of having highly compatible globalization mode across all platform we support we need to look into bringing ICU to ios/tvOS as well.

> System.Globalization.Invariant mode still need to work without any ICU dependency. How we are actually going to slice the actual data will be defined later based on the experience and existing ICU tooling.

Reviewer: >actual data will be defined later based on the experience  [](start = 122, length = 58)

can you add TBD section for this part to describe later the slicing specs?

We are as expected not the first framework to hit this problem. Swift framework is dealing with same problems and they seem to prefer a solution which drop ICU dependency but that still does not look like final [decision](https://forums.swift.org/t/icu-usage-in-swift/20473).

### WebAssembly
The intention is set invariant globalization mode as the default mode for everyone for WebAssembly but at the same time allow a fallback to full globalization support when needed. This will come with additional cost but on the web platform we have little bit more flexibility how to deploy and share the data.

> ICU library consists from code and data. What we need to explore is if there is reasonable way to package the ICU data and store them in the cloud as a publicly available data blobs. This could give us flexibility to support full range of ICU data on demand if sliced into small data sets. We’ll still need to ship icu implementation but that could be also bound dynamically using dynamic linking.

Reviewer: >but that could be also bound dynamically using dynamic linking [](start = 334, length = 62)

Windows is going to ship ICU packages for Windows which is tweaked a little to match MS bar. do you think ICU iOS packages need to do the same?

Reviewer 2: @tarekgh What kind of tweaking is applied to Windows ICU?

Author: I don't believe it'll need to be tweaked.

## ICU Packaging
Due to lack of ICU on all platform we intend to support for .NET5 we need to look into how to package it for following RIDs

> * ios-x64

Reviewer: >os-x64 [](start = 3, length = 6)

if we are going to have ICU package for iOS. do you think the exact same package can work for tvos? or we have to recompile it to make it work there?

Reviewer 7: We need to compile the library against tvOS too, it's a different platform and iOS binaries don't work there.

* ios-arm
* ios-arm64
* tvos-x64
* tvos-arm64
* webassembly (browser mode)

> A default build of ICU normally results in over 16 MB of data, and a substantial amount of object code.

Reviewer 5: 16MB (data only) is more than twice the size of a basic iOS/tvOS application

Reviewer 4: Please read slicing section

Reviewer 5: The only places that mention slicing speaks in conditional and don't offer any numbers.

I'll check how big is a current XI app when all I18N assemblies are included. I realize it's not the same, feature/completion wise, but the extraneous data/code is not needed for everyone (but the size impact will be).

Reviewer 5: iPhone Release (arm64 only) Builds Application Size (XI master)

* No i18n assembly: 5.1 MB
* All i18n assemblies: 6.9 MB 

all = I18N.dll, I18N.West.dll, I18NRare.dll, I18N.Other.dll, I18N.MidEast.dll, I18N.CJK.dll

Reviewer 4: Just for completeness you also have to add runtime globalization data and code

> The packages have to be split to implementation and data part and ideally the data parts will be also sliced in the way that the developers can select only relevant data for their app if they for example don’t target Chinese market.

Reviewer 2: Would it make sense to statically link the ICU and the System.Globalization shim together, so that we only carry implementation for the the ICU APIs that the shim needs?

Reviewer 8: Would that break app local scenarios? This way if people want to upgrade to a newer ICU, they couldn't because we carry the implementations that we built against, no?

Reviewer 2: This paragraph is proposing to have our own custom build of ICU. There would not be a newer version unless we build it.

Reviewer 8: Ah I see. I thought you were proposing that for all scenarios.

Author: I don't know what the implications are, if any, behind that.  I'm sure it's something we could determine during when implementing.

### Data Library Customizer 
There is a [tool](http://www.icu-project.org/docs/demo/datacustom_help.html) built by ICU which allow customizing what kind of data will be included in the final data set. We need to explore if that something we could leverage either for the data islands or even as something using during build process.

Additional documentation about the tool is [here](https://github.com/unicode-org/icu/blob/master/docs/userguide/icu_data/buildtool.md)

## App Local ICU for Windows

> Windows teams is working on packaging ICU for .NET needs. The package will be available as NuGet package and we could explore to use similar mechanism for iOS/tvOS packaging.

Reviewer: >The package will be available as NuGet package  [](start = 58, length = 47)

it worth to mentioned Windows ICU package is including the whole data and will not include any customized data packages as required for iOS/tvOS

## .NET Core ICU dependencies
All of the existing ICU dependencies in .NET Core today are located inside System.Private.CoreLib. The .NET Core does not install ICU and relies on the unmanaged implementation to be available on the system.

The existing implementation requires ICU version is [50](https://github.com/dotnet/runtime/blob/4480bdbe66c55caceb64fc3477010435ba70c4ec/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L60) or higher.

The implementation files which reference globalization interop are:

* [System.Private.CoreLib/src/System/Globalization/IdnMapping.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/IdnMapping.Unix.cs)
* [System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs)
* [System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs)

> * [System.Private.CoreLib/src/System/Globalization/TextInfo.Unix.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.Unix.cs)

Reviewer 8: These files were renamed to .Icu.cs instead of .Unix.cs on my PR to add support for Windows.

Reviewer 7: In general it'd be good not to link to master but an explicit SHA so these links don't get stale.

Consult this [issue](https://github.com/android/ndk/issues/548#issuecomment-395561629) for detailed commentary on the challenges of making ICU an NDK API.

### What Does This Mean For Us?
The good news is that p-invoking from managed code works.  Assuming `libicuuc` continues to be white listed, this should not be a problem.  It would be wise if we explored using the `ICU4J`  API subset that’s exposed in the event they decide to tighten things up.  

To summarize:

* We can p-invoke into the system `libicuuc`.  Globalization for .net5 should work normally
* Android 4.3 (API level 18) started bundling ICU 50 and our minimum version requirement is satisfied. 
* OEM’s can potentially make life difficult by altering locales / globalization features
* We should explore trying to interop with the `ICU4J` API and determine what the implications, if any are.

> ### iOS/tvOS

Reviewer 5: That section should also include **watchOS**. 

watchOS it has application size limitations (50MB) which makes brining additional dependencies very hard.

Author: It's my understanding that watchOS is not included in .net5.  Would it make sense to consider now?

Reviewer 5: care to link me to that decision ? because it's right there in https://devblogs.microsoft.com/dotnet/introducing-net-5/

Reviewer 9: @steveisok (and @marek-safar) - Can we please get confirmation on what platforms / archs / features are "not included in .net 5" as of right now. 

I'd like to get a handle on it, and be able to get it in front of customers before release this fall.

Author: For some reason I'm not getting notifications, so my apologies.  I had thought that was official, but that's clearly my assumption may be incorrect.  @marek-safar ?

Reviewer 10: @steveisok and @marek-safar  Did you get any clarification on WatchOS support?  I'm building a new app using Xamarin and WatchOS and I don't want to waste my time if support will be dropped.  

I have two apps in the store ([here](https://apps.apple.com/us/app/voice-in-a-can/id1339696720?platform=appleWatch) and [here](https://apps.apple.com/us/app/kanbann/id1480933645#?platform=appleWatch)) currently using Xamarin and WatchOS which would be orphaned if you drop support.

Reviewer 4: @DamianMehers Xamarin won't drop WatchOS support. Here we are discussing the priorities of how to migrate that to .NET 5/6/7.

Reviewer 10: Understood - my concern is that if it doesn't make it to .NET 5 then I assume eventually support will disappear.

There is no build in support for ICU on apple mobile platforms. For our purposes of having highly compatible globalization mode across all platform we support we need to look into bringing ICU to ios/tvOS as well.

System.Globalization.Invariant mode still need to work without any ICU dependency. How we are actually going to slice the actual data will be defined later based on the experience and existing ICU tooling.

> We are as expected not the first framework to hit this problem. Swift framework is dealing with same problems and they seem to prefer a solution which drop ICU dependency but that still does not look like final [decision](https://forums.swift.org/t/icu-usage-in-swift/20473).

Reviewer 5: That's a bad example. The issue is tagged for `windows`, `linux` and `android`, all non-Apple platforms, so any _decision_ is not applicable (in this section, at least).

This is not a problem (for Swift) on Apple platforms because ICU is part of the OS. Swift is (now) a system framework and so can use ICU directly.

However ICU is not (directly) accessible to developers. We can only use the Apple provided API (that are using ICU internally).

Author: Does the Apple API give us enough coverage?  It was not clear to me if we could / want to use it.

Reviewer 5: Coverage ? yes
Enough ? that's an elastic notion ;-)

Sorry but it's not my area of expertise :( so I'm not sure if it's a super or subset of what .net provide.

Author: I would still suggest we stay with keeping ICU as an external dependency.  Perhaps as we move along we can test the viability of using the underlying API?  This seems fairly new to me.

### WebAssembly
The intention is set invariant globalization mode as the default mode for everyone for WebAssembly but at the same time allow a fallback to full globalization support when needed. This will come with additional cost but on the web platform we have little bit more flexibility how to deploy and share the data.

> ICU library consists from code and data. What we need to explore is if there is reasonable way to package the ICU data and store them in the cloud as a publicly available data blobs. This could give us flexibility to support full range of ICU data on demand if sliced into small data sets. We’ll still need to ship icu implementation but that could be also bound dynamically using dynamic linking.

Reviewer 8: Currently libicuuc and libicui18n are dynamically linked to libicudata so we would need to download and make libicudata available before we load ICU.

# Globalization Support in .NET 5

## Overview
Globalization in the .NET Core space has been handled by taking a dependency on the [icu4c](http://site.icu-project.org/home) library and p-invoking into it from managed code.  This model has been successful for desktop and server modes mostly because they can afford to take on the disk space ICU requires. However, in the size restricted workloads, this is quite a different story as there are differences in platform support and tolerance of taking on a significantly sized dependency.

The purpose of this document is to describe what globalization support, if any, Android, iOS/tvOS, and WebAssembly have and what options there are should ICU not suffice.

## .NET Core ICU dependencies

> All of the existing ICU dependencies in .NET Core today are located inside System.Private.CoreLib. The .NET Core does not install ICU and relies on the unmanaged implementation to be available on the system.

Reviewer 11: ```suggestion
All of the existing ICU dependencies in .NET Core today are located inside System.Private.CoreLib. .NET Core does not install ICU and relies on the unmanaged implementation to be available on the system.
```

> The existing implementation requires ICU version is [50](https://github.com/dotnet/runtime/blob/4480bdbe66c55caceb64fc3477010435ba70c4ec/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L60) or higher.

Reviewer 11: ```suggestion
The existing implementation requires that the ICU version is [50](https://github.com/dotnet/runtime/blob/4480bdbe66c55caceb64fc3477010435ba70c4ec/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L60) or higher.
```

* [System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs](https://github.com/dotnet/runtime/tree/87f391001eca716a8db896f5c3855d33fe30aca8/src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs)
* [System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Icu.cs](https://github.com/dotnet/runtime/tree/87f391001eca716a8db896f5c3855d33fe30aca8/src/libraries/System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Icu.cs)
* [https://github.com/dotnet/runtime/blob/88cc71b26ba1eb430aaeeebddcd5675b3009f2ee/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L127](https://github.com/dotnet/runtime/blob/88cc71b26ba1eb430aaeeebddcd5675b3009f2ee/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L127)

Existing test coverage checks the implementation which is available. It does not have any expected list of data included therefore does not check whether same set of data is available everywhere.

## .NET Globalization Invariant Mode

> There is already support for the globalization [invariant mode](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md) since .NET Core 2.0. This will be the most size effective globalization mode for .NET5 which can be set by the developers. We are going to use existing logic and tweak it where necessary to allow IL Linker to remove any globalization code from the final app when running in invariant mode.

Reviewer 11: ```suggestion
There is already support for the globalization [invariant mode](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md) since .NET Core 2.0. This will be the most size effective globalization mode for .NET 5 which can be set by the developers. We are going to use existing logic and tweak it where necessary to allow IL Linker to remove any globalization code from the final app when running in invariant mode.
```

### Enhanced Invariant Mode
In order to cast a wider net for applications that can use invariant mode, we will incorporate the simple ICU casing tables into dotnet/runtime.  This allows us to move beyond ASCII when it comes to casing comparisons at a very small (3Kb) size cost. 

> See the enhanced invariant mode [https://github.com/dotnet/runtime/issues/30960](proposal) for more detail.

Reviewer 11: ```suggestion
See the enhanced invariant mode [proposal](https://github.com/dotnet/runtime/issues/30960) for more detail.
```

## Size Sensitive Platforms

> Existing Xamarin platforms allow customers to choose which set of collation data they would like to add to their apps if any. This still makes sense for .NET5 as on some of the platform ICU dependency is not available from the system and we’ll need to bundle it to the final app. We could make this available to everyone but it’s most likely we will focus only on platforms which are size sensitive in .NET5 time frame.

Reviewer 11: ```suggestion
Existing Xamarin platforms allow customers to choose which set of collation data they would like to add to their apps, if any. This still makes sense for .NET 5, as on some platforms the ICU dependency is not available from the system, and we’ll need to bundle it to the final app. We could make this available to everyone, but it’s most likely we will focus only on platforms which are size-sensitive in the .NET 5 time frame.
```

In API level 24 and higher, the platform white lists its version of `libicuuc` and also includes a subset of [ICU4J](https://developer.android.com/guide/topics/resources/internationalization#relation) API’s.  It’s important to note that [ICU4C](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/) (accessing `libicuuc` directly) is not part of the Android NDK.  What this means is that OEM’s are not required to follow any type of rules and they could limit / alter the data files as they see fit.  The potential for undefined behavior and additional testing on our part is very real. 

Consult this [issue](https://github.com/android/ndk/issues/548#issuecomment-395561629) for detailed commentary on the challenges of making ICU an NDK API.

### What Does This Mean For Us?
The good news is that p-invoking from managed code works.  Assuming `libicuuc` continues to be white listed, this should not be a problem.  It would be wise if we explored using the `ICU4J`  API subset that’s exposed in the event they decide to tighten things up.  

To summarize:

> * We can p-invoke into the system `libicuuc`.  Globalization for .net5 should work normally

Reviewer 11: ```suggestion
* We can p-invoke into the system `libicuuc`.  Globalization for .NET 5 should work normally
```

* Android 4.3 (API level 18) started bundling ICU 50 and our minimum version requirement is satisfied. 
* OEM’s can potentially make life difficult by altering locales / globalization features
* We should explore trying to interop with the `ICU4J` API and determine what the implications, if any are.

### iOS/tvOS

> There is no build in support for ICU on apple mobile platforms. For our purposes of having highly compatible globalization mode across all platform we support we need to look into bringing ICU to ios/tvOS as well.

Reviewer 11: ```suggestion
There is no built-in support for ICU on Apple mobile platforms. For our purposes of having highly compatible globalization mode across all platforms we support, we need to look into bringing ICU to ios/tvOS as well.
```

> System.Globalization.Invariant mode still need to work without any ICU dependency. How we are actually going to slice the actual data will be defined later based on the experience and existing ICU tooling.

Reviewer 11: ```suggestion
System.Globalization.Invariant mode still needs to work without any ICU dependency. How we are going to slice the actual data will be defined later based on the experience and existing ICU tooling.
```

> We are as expected not the first framework to hit this problem. Swift framework is dealing with same problems and they seem to prefer a solution which drop ICU dependency but that still does not look like final [decision](https://forums.swift.org/t/icu-usage-in-swift/20473).

Reviewer 11: ```suggestion
We are not the first framework to hit this problem. Swift framework is dealing with same problems, and they seem to prefer a solution which drops the ICU dependency, but that still does not look like final [decision](https://forums.swift.org/t/icu-usage-in-swift/20473).
```

### WebAssembly

> The intention is set invariant globalization mode as the default mode for everyone for WebAssembly but at the same time allow a fallback to full globalization support when needed. This will come with additional cost but on the web platform we have little bit more flexibility how to deploy and share the data.

Reviewer 11: ```suggestion
The intention is to set invariant globalization mode as the default mode for everyone for WebAssembly, but at the same time allow a fallback to full globalization support when needed. This will come with additional cost, but on the web platform we have a little bit more flexibility how to deploy and share the data.
```

> ICU library consists from code and data. What we need to explore is if there is reasonable way to package the ICU data and store them in the cloud as a publicly available data blobs. This could give us flexibility to support full range of ICU data on demand if sliced into small data sets. We’ll still need to ship icu implementation but that could be also bound dynamically using dynamic linking. 

Reviewer 11: ```suggestion
The ICU library consists from code and data. We need to explore if there is a reasonable way to package the ICU data and store them in the cloud as publicly available data blobs. This could give us flexibility to support the full range of ICU data on demand if sliced into small data sets. We’d still need to ship the ICU implementation, but that also could be bound dynamically using dynamic linking. 
```

From our earlier ICU work, we do know that libicuuc and libicui18n are dynamically linked to libicudata. We would need to download and make libicudata available before loading ICU. 

## ICU Packaging

> Due to lack of ICU on all platform we intend to support for .NET5 we need to look into how to package it for following RIDs

Reviewer 11: ```suggestion
Due to lack of ICU on all platform we intend to support for .NET 5, we need to look into how to package it for following the RIDs:
```

* ios-x64
* ios-arm
* ios-arm64
* tvos-x64
* tvos-arm64
* webassembly (browser mode)

A default build of ICU normally results in over 16 MB of data, and a substantial amount of object code.

> The packages have to be split to implementation and data part and ideally the data parts will be also sliced in the way that the developers can select only relevant data for their app if they for example don’t target Chinese market.

Reviewer 11: ```suggestion
The packages have to be split into implementation and data parts, and ideally the data parts also will be sliced in a way that the developers can select only the relevant data for their app if they only target specific markets.
```

### Data Library Customizer (TBD)

> There is a [tool](http://www.icu-project.org/docs/demo/datacustom_help.html) built by ICU which allow customizing what kind of data will be included in the final data set. We need to explore if that something we could leverage either for the data islands or even as something using during build process.

Reviewer 11: ```suggestion
There is a [tool](http://www.icu-project.org/docs/demo/datacustom_help.html) built by ICU which allows customizing what kind of data will be included in the final data set. We need to explore if that is something we could leverage, either for the data islands or even as something used during the build process.
```

Additional documentation about the tool is [here](https://github.com/unicode-org/icu/blob/master/docs/userguide/icu_data/buildtool.md)

## App Local ICU for Windows

> Windows teams is working on packaging ICU for .NET needs. It will be available as NuGet package and we could explore to use a similar mechanism for iOS/tvOS packaging (although Windows will contain the whole data file). 

Reviewer 11: ```suggestion
The Windows teams is working on packaging ICU for .NET needs. It will be available as a NuGet package, and we could explore using a similar mechanism for iOS/tvOS packaging (although Windows will contain the whole data file). 
```
